### ISSUE TITLE
[Metrics] Feedback on MetricManager API

### ISSUE THREAD

**Issue by AlexBulankou** [Microsoft/ApplicationInsights-dotnet#439]
Will the following code result into pre-aggregated  ``test`` metric sent:
```
                new MetricManager(new TelemetryClient()).CreateMetric("test").Track(1);
                new MetricManager(new TelemetryClient()).CreateMetric("test").Track(2);
```
or 2 separate MetricTelemetry items? As a customer I would like API to suggest the optimal approach to me and from this perspective having to initialize new MetricManager object and having to Create and keep around my instance of Metric appears too much work. I already learnt how to do this:
```
new TelemetryClient().TrackMetric("test", 1);
```
Additional confusion is that ``Metric`` is different from ``MetricTelemetry`` that we all know. 
Can we add ``TrackAggregated`` without having to create MetricManager and keep the reference to metric.
```
new TelemetryClient().TrackMetricAggregated("test", 1) 
```

*Suggestion:* Add ``TrackAggregatedMetric`` API on TelemetryClient for more intuitive use and to not require to create instance of MetricManager and Metric.

**Comment by SergeyKanzhelev** [comment]
cc: @vitalyf007 @olegsych @macrogreg 

We discussed multiple implementations of this API. this is a good feedback and I'd appreciate if you'll provide a specific example when you had such confusion rather than opinion.

**Comment by AlexBulankou** [comment]
Yesterday I upgraded my app to 2.3 Beta2 (from 2.2 stable) and, as a customer would, saw warnings saying that `TrackMetric(name, value)` is now obsolete. I understand that metrics are now aggregated on the client, however my first thought was, why do I have to do anything different, please go ahead and aggregate my metric on the client, not the server. Ah, I see,  I have to pick dimensions now. Well, can you pick them for me? Sure, I realize, this may be not so straightforward. So I can pick dimensions and even start using different API, but why am I asked to change my program structure, create `MetricManager` and create and retain Metric objects. I think we could benefit from API that would be make upgrade easier:

- My first preference would be: no changes to `TrackMetric` but now there's client-side aggregation. Pick important dimensions for me.
- My second preference: modify `TrackMetric(name, ...)` and `TrackMetric(MetricTelemetry, ...)` APIs so they support aggregated option
- My third preference: introduce `TrackMetricAggregated` that supports same usage model as `TrackMetric`

**Comment by SergeyKanzhelev** [comment]
@vitalyf007 where are you on this issue https://github.com/Microsoft/ApplicationInsights-dotnet/issues/388? We need a proper documentation on recommended approach to transition from `TrackMetric` to `MetricManager`.

I'm renaming this issue to "feedback on metircManager API" and will post more issues that I hit

**Comment by SergeyKanzhelev** [comment]
More feedback after implementing metrics aggregation for Exception Statistics:

1. I want to set `.Context.Operation.Name` for a aggregated metric as @AlexBulankou suggested https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/327 MetricManager doesn't allow to do it. Should we implement this capability or keep it as is?
2. I want to set `.Context.Operation.SyntheticSource` on aggregated metric to indicate another "predefined" context property. Same problem as above
3. I am creating a new metric object on every `Track` as I need to specify dimensions values on creation, not on `Track`. This makes code look bad and it is not optimal from performance standpoint

**Comment by SergeyKanzhelev** [comment]
What do you think of moving `MetricManager` API under `Experimental` or `Implementation` namespace before releasing 2.3 stable and clearly state in docs that this API is subject to change?

Current API stuck in the middle between simple `InterlockedIncrement(metricValue)` and customer manages the lifecycle and processing metric thru the pipeline `TrackMetric() --> call Initializers --> aggregate by all properties`. If it causes confusion like @AlexBulankou mentioned - we may want to postpone marking it stable.

**Comment by zakimaksyutov** [comment]
I agree with @AlexBulankou that we should not depreciate old API. Customers already aggregate and they way they do it might not be easy/possible to migrate to our new API. I think TrackMetric makes sense - we should consider as extensibility point - it allows to plug any other metric APIs.

Re: complexity. There are two different scenarios/usage patterns:
1) Low volume app (<100 RPS/node) - for such apps I'd like to use as simple API as possible. TrackMetricAggregated from @AlexBulankou suggestion would work for me well.
2) High volume app (> 2000 RPS/node) - here we should provide lock-free implementation essentially based on InterlockedIncrement. In order to avoid performance hit as a customer I'd be willing to do metric object management myself.

**Comment by olegsych** [comment]
As I recall, we wanted to deprecate `TrackMetric` because making it aggregate metrics instead of sending them immediately is a breaking change that requires a major version upgrade. Another reason for deprecating it was to avoid confusion regarding whether the standard `TelemetryClient.Context` properties, like `TelemetryContext.Operation` become dimensions for aggregation. We intentionally decided that context properties are _not_ aggregation dimensions and making `TrackMetric` aggregate using `TelemetryClient.Context.Properties` would be confusing.

**Comment by AlexBulankou** [comment]
@olegsych I think I disagree. I think we **should** be making standard set of context properties aggregation dimensions. If this is not feasible due to memory reasons, we should be making it for a subset of context properties, i.e. start with standard properties and then decide which ones we cut from metric aggregation, or let advanced users decide which ones to cut. If we don't do it, the users will be forced to reinvent standard dimensions causing pollution on the UI and inability to use our APM drilldown experience (consider Sergey's exceptions statistics with ``method`` and ``operation``).  Also how are we planning to implement metric aggregation for standard metrics in 2.4Beta 4 if not by including standard properties? 
If we absolutely feel that we should not be including standard properties by default I would advocate API override that makes it very simple. For example:
```
myTelemetryClient.TrackMetricAggregated("EventHubConnectionAttempts", attempts, includeStanardDimensions:true, {"EventHubStamp", ehStamp});
```

**Comment by olegsych** [comment]
@AlexBulankou I understand that we've encountered a challenge when implementing tracking of aggregated metrics in the SDK (Microsoft/ApplicationInsights-dotnet-server#327). Is it possible to solve this problem in the implementation of the SDK as opposed to making a high-risk change in the public API at this stage?

**Comment by SergeyKanzhelev** [comment]
@olegsych this API is in beta only. So we can still gather feedback and make changes if needed. I asked @zakimaksyutov to onboard his service and provide more real life feedback.

**Comment by macrogreg** [comment]
Putting things into an experimental namespace literally makes sure that a breaking change will occur down the line. I would much rather use the intended namespace for Beta versions of the SDK (xxx.Metrics ?) and potentially not include the APIs into the final release, unless we are sure that we are happy with the API shape.

However, I agree with @AlexBulankou that something like 
    new TelemetryClient().TrackMetricAggregated("test", 1) 
would be helpful.

**Comment by SergeyKanzhelev** [comment]
@macrogreg we plan to ship 2.3 stable soon. We need this API for exceptions statistics and your work on standard metrics. As well as internal testing and feedback like above. 

So to not lock on API and still be able to ship the stable version - I propose to move it under `Experimental`.

Moving to separate NuGet wouldn't work as it will require taking dependency on that NuGet from other NuGets like `ServerChannel` and `Windows`.

**Comment by macrogreg** [comment]
At the API as it was shipped in 2.4-beta1 is no longer shipped and being updated, I am closing this issue for now.
